### PR TITLE
EQL: Fix bug in skipping window

### PR DESCRIPTION
**Pull request by costin**
Fix NPE when closing a sequence group
Increase randomization in testing
Make maxspan inclusive (up to and equal to value vs just up to).

### CODE REVIEW

**Review by astefan — APPROVED** [review]
LGTM

**Comment by costin** [comment]
@elasticmachine update branch

File: client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchRequest.java [hunk 1]
```diff
@@ -191,8 +191,8 @@ public int fetchSize() {
         return this.fetchSize;
     }
 
-    public EqlSearchRequest fetchSize(int size) {
-        this.fetchSize = size;
+    public EqlSearchRequest fetchSize(int fetchSize) {
```
Comment (costin) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchRequest.java:194]: @matriv Included your feedback from the previous PR.

File: x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/CommonEqlActionTestCase.java [hunk 2]
```diff
@@ -146,7 +146,7 @@ protected EqlSearchResponse runQuery(String index, String query, boolean isCaseS
         request.tiebreakerField("event.sequence");
         // some queries return more than 10 results
         request.size(50);
-        request.fetchSize(2);
+        request.fetchSize(randomIntBetween(2, 50));
```
Comment (costin) [inline x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/CommonEqlActionTestCase.java:149]: @matriv see this one as well. It unraveled a bug - when the fetchSize was 5, the results were different due to us skipping a window.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceStateMachine.java [hunk 3]
```diff
@@ -82,19 +83,19 @@ public void match(int stage, SequenceKey key, Ordinal ordinal, SearchHit hit) {
...
         group.trim(before.v2() + 1);
+
+        // remove the frame and keys early (as the key space is large)
+        if (group.isEmpty()) {
+            stageToKeys.keys(previousStage).remove(key);
+        }
         
         // check maxspan before continuing the sequence
-        if (maxSpanInMillis > 0 && (ordinal.timestamp() - sequence.startTimestamp() >= maxSpanInMillis)) {
+        if (maxSpanInMillis > 0 && (ordinal.timestamp() - sequence.startTimestamp() > maxSpanInMillis)) {
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceStateMachine.java:93]: Made maxspan to be inclusive.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceStateMachine.java [hunk 4]
```diff
@@ -117,7 +118,27 @@ public boolean reachedLimit() {
...
+    /**
+     * Checks whether the rest of the stages have in-flight data.
+     * This method is called when a query returns no data meaning
+     * sequences on previous stages cannot match this window (since there's no new data).
+     * However sequences on higher stages can, hence this check to know whether
+     * it's possible to advance the window early.
+     */
     public boolean hasCandidates(int stage) {
-        return stage < completionStage && stageToKeys.keys(stage).isEmpty() == false;
+        for (int i = stage; i < completionStage; i++) {
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceStateMachine.java:129]: This turned out to be causing data skipping by ignoring potential matches for following stages.

**Review by matriv — APPROVED** [review]
LGTM!

**Pull request merged**